File: lib/Json.js
import fs from 'fs' ;
import path from 'path';

/**
 * Class for managing a simple JSON database.
 * 
 * This class provides basic operations such as create, read, 
 * update, and delete data in JSON format.
 */
class JsonDB {
    /**
     * Initialize JSON database.
     * 
     * @param {string} filePath - Path to the JSON file to be used as database
     */
    constructor(filePath) {
        this.filePath = filePath;
        this.data = {};
        
        // Create directory if it doesn't exist
        const directory = path.dirname(filePath);
        if (directory && !fs.existsSync(directory)) {
            fs.mkdirSync(directory, { recursive: true });
        }
        
        // Create empty file if it doesn't exist
        if (!fs.existsSync(filePath)) {
            this._saveToFile({});
        }
        
        // Load data from file
        this._loadFromFile();
    }
    
    /**
     * Load data from JSON file to memory.
     * @private
     */
    _loadFromFile() {
        try {
            const content = fs.readFileSync(this.filePath, 'utf-8').trim();
            if (content) {
                this.data = JSON.parse(content);
            } else {
                this.data = {};
            }
        } catch (error) {
            console.log(`Warning: Failed to load JSON file. Using empty database. Error: ${error.message}`);
            this.data = {};
        }
    }
    
    /**
     * Save data to JSON file.
     * 
     * @param {Object} [data] - Data to be saved. If null, use this.data
     * @private
     */
    _saveToFile(data = null) {
        const saveData = data !== null ? data : this.data;
        try {
            fs.writeFileSync(this.filePath, JSON.stringify(saveData, null, 2), 'utf-8');
        } catch (error) {
            console.log(`Error: Failed to save to file. ${error.message}`);
        }
    }
    
    /**
     * Create a new table in the database.
     * 
     * @param {string} tableName - Name of the table to be created
     * @param {*} [initialData] - Initial data for the table (optional)
     * @returns {boolean} True if successful, false if table already exists
     */
    createTable(tableName, initialData = null) {
        if (tableName in this.data) {
            console.log(`Warning: Table '${tableName}' already exists.`);
            return false;
        }
        
        // If no initial data, create as empty array
        this.data[tableName] = initialData !== null ? initialData : [];
        this._saveToFile();
        return true;
    }
    
    /**
     * Add data to table.
     * 
     * @param {string} tableName - Table name
     * @param {*} data - Data to be added
     * @returns {boolean} True if successful
     */
    insertData(tableName, data) {
        if (!(tableName in this.data)) {
            console.log(`Warning: Table '${tableName}' not found. Creating new table.`);
            this.createTable(tableName);
        }
        
        // If table is an array, push data
        if (Array.isArray(this.data[tableName])) {
            this.data[tableName].push(data);
        }
        // If table is object and data is also object, merge
        else if (typeof this.data[tableName] === 'object' && 
                 this.data[tableName] !== null && 
                 typeof data === 'object' && 
                 data !== null &&
                 !Array.isArray(this.data[tableName]) &&
                 !Array.isArray(data)) {
            Object.assign(this.data[tableName], data);
        }
        // If table is single value, convert to array
        else {
            const oldData = this.data[tableName];
            this.data[tableName] = [oldData, data];
        }
        
        this._saveToFile();
        return true;
    }
    
    /**
     * Update data in table.
     * 
     * @param {string} tableName - Table name
     * @param {Function} [condition] - Function to determine which data to update
     * @param {*} [newData] - New data
     * @returns {boolean} True if successful
     */
    updateData(tableName, condition = null, newData = null) {
        if (!(tableName in this.data)) {
            console.log(`Error: Table '${tableName}' not found.`);
            return false;
        }
        
        // If no condition, replace entire data
        if (condition === null) {
            this.data[tableName] = newData;
        } else {
            // For array, update items that meet condition
            if (Array.isArray(this.data[tableName])) {
                for (let i = 0; i < this.data[tableName].length; i++) {
                    if (condition(this.data[tableName][i])) {
                        this.data[tableName][i] = newData;
                    }
                }
            }
        }
        
        this._saveToFile();
        return true;
    }
    
    /**
     * Delete data from table.
     * 
     * @param {string} tableName - Table name
     * @param {Function} [condition] - Function to determine which data to delete
     * @param {boolean} [deleteAll=false] - If true, delete all data in table
     * @returns {boolean} True if successful
     */
    deleteData(tableName, condition = null, deleteAll = false) {
        if (!(tableName in this.data)) {
            console.log(`Error: Table '${tableName}' not found.`);
            return false;
        }
        
        if (deleteAll) {
            if (Array.isArray(this.data[tableName])) {
                this.data[tableName] = [];
            } else if (typeof this.data[tableName] === 'object' && this.data[tableName] !== null) {
                this.data[tableName] = {};
            } else {
                this.data[tableName] = null;
            }
        } else if (condition) {
            if (Array.isArray(this.data[tableName])) {
                this.data[tableName] = this.data[tableName].filter(item => !condition(item));
            }
        }
        
        this._saveToFile();
        return true;
    }
    
    /**
     * Retrieve data from table or entire database.
     * 
     * @param {string} [tableName] - Table name (optional)
     * @returns {*} Data from table or entire database
     */
    getData(tableName = null) {
        if (tableName === null) {
            return this.data;
        }
        
        if (!(tableName in this.data)) {
            console.log(`Warning: Table '${tableName}' not found.`);
            return null;
        }
        
        return this.data[tableName];
    }
    
    /**
     * Display data information in readable format.
     * 
     * @param {string} [tableName] - Table name (optional)
     */
    showData(tableName = null) {
        if (tableName === null) {
            console.log("=== Entire Database ===");
            for (const [key, value] of Object.entries(this.data)) {
                console.log(`Table: ${key}`);
                console.log(`Type: ${this._getType(value)}`);
                if (Array.isArray(value) || typeof value === 'object' || typeof value === 'string') {
                    console.log(`Length: ${this._getLength(value)}`);
                }
                console.log(`Data: ${JSON.stringify(value, null, 2)}`);
                console.log("-".repeat(40));
            }
        } else {
            if (!(tableName in this.data)) {
                console.log(`Table '${tableName}' not found.`);
                return;
            }
            
            const data = this.data[tableName];
            console.log(`=== Table: ${tableName} ===`);
            console.log(`Type: ${this._getType(data)}`);
            if (Array.isArray(data) || typeof data === 'object' || typeof data === 'string') {
                console.log(`Length: ${this._getLength(data)}`);
            }
            console.log(`Data: ${JSON.stringify(data, null, 2)}`);
        }
    }
    
    /**
     * Return list of all table names.
     * 
     * @returns {string[]} Array of table names
     */
    listTables() {
        return Object.keys(this.data);
    }
    
    /**
     * Check if table exists.
     * 
     * @param {string} tableName - Table name
     * @returns {boolean} True if table exists
     */
    tableExists(tableName) {
        return tableName in this.data;
    }
    
    /**
     * Get detailed information about table.
     * 
     * @param {string} tableName - Table name
     * @returns {Object} Object with table information
     */
    getTableInfo(tableName) {
        if (!(tableName in this.data)) {
            return { exists: false };
        }
        
        const data = this.data[tableName];
        return {
            exists: true,
            name: tableName,
            type: this._getType(data),
            length: this._getLength(data),
            data: data
        };
    }
    
    /**
     * Get type of value.
     * 
     * @param {*} value - Value to check type
     * @returns {string} Type name
     * @private
     */
    _getType(value) {
        if (Array.isArray(value)) return 'Array';
        if (value === null) return 'null';
        return typeof value;
    }
    
    /**
     * Get length of value if applicable.
     * 
     * @param {*} value - Value to check length
     * @returns {number} Length or 0
     * @private
     */
    _getLength(value) {
        if (Array.isArray(value) || typeof value === 'string') {
            return value.length;
        }
        if (typeof value === 'object' && value !== null) {
            return Object.keys(value).length;
        }
        return 0;
    }
}

export { JsonDB };

File: Json.py
import json
import os
from typing import Any, Dict, List, Union, Optional

class JsonDB:
    """
    Class for managing a simple JSON database.
    
    This class provides basic operations such as create, read, 
    update, and delete data in JSON format.
    """
    
    def __init__(self, file_path: str):
        """
        Initialize JSON database.
        
        Args:
            file_path (str): Path to the JSON file to be used as database
        """
        self.file_path = file_path
        self.data = {}
        
        # Create directory if it doesn't exist
        directory = os.path.dirname(file_path)
        if directory and not os.path.exists(directory):
            os.makedirs(directory)
        
        # Create empty file if it doesn't exist
        if not os.path.exists(file_path):
            self._save_to_file({})
        
        # Load data from file
        self._load_from_file()
    
    def _load_from_file(self) -> None:
        """Load data from JSON file to memory."""
        try:
            with open(self.file_path, 'r', encoding='utf-8') as file:
                content = file.read().strip()
                if content:
                    self.data = json.loads(content)
                else:
                    self.data = {}
        except (json.JSONDecodeError, FileNotFoundError) as e:
            print(f"Warning: Failed to load JSON file. Using empty database. Error: {e}")
            self.data = {}
    
    def _save_to_file(self, data: Dict[str, Any] = None) -> None:
        """
        Save data to JSON file.
        
        Args:
            data (Dict): Data to be saved. If None, use self.data
        """
        save_data = data if data is not None else self.data
        try:
            with open(self.file_path, 'w', encoding='utf-8') as file:
                json.dump(save_data, file, ensure_ascii=False, indent=2)
        except Exception as e:
            print(f"Error: Failed to save to file. {e}")
    
    def create_table(self, table_name: str, initial_data: Any = None) -> bool:
        """
        Create a new table in the database.
        
        Args:
            table_name (str): Name of the table to be created
            initial_data (Any): Initial data for the table (optional)
            
        Returns:
            bool: True if successful, False if table already exists
        """
        if table_name in self.data:
            print(f"Warning: Table '{table_name}' already exists.")
            return False
        
        # If no initial data, create as empty list
        self.data[table_name] = initial_data if initial_data is not None else []
        self._save_to_file()
        return True
    
    def insert_data(self, table_name: str, data: Any) -> bool:
        """
        Add data to table.
        
        Args:
            table_name (str): Table name
            data (Any): Data to be added
            
        Returns:
            bool: True if successful
        """
        if table_name not in self.data:
            print(f"Warning: Table '{table_name}' not found. Creating new table.")
            self.create_table(table_name)
        
        # If table is a list, append data
        if isinstance(self.data[table_name], list):
            self.data[table_name].append(data)
        # If table is dict and data is also dict, update
        elif isinstance(self.data[table_name], dict) and isinstance(data, dict):
            self.data[table_name].update(data)
        # If table is single value, convert to list
        else:
            old_data = self.data[table_name]
            self.data[table_name] = [old_data, data]
        
        self._save_to_file()
        return True
    
    def update_data(self, table_name: str, condition: callable = None, new_data: Any = None) -> bool:
        """
        Update data in table.
        
        Args:
            table_name (str): Table name
            condition (callable): Function to determine which data to update
            new_data (Any): New data
            
        Returns:
            bool: True if successful
        """
        if table_name not in self.data:
            print(f"Error: Table '{table_name}' not found.")
            return False
        
        # If no condition, replace entire data
        if condition is None:
            self.data[table_name] = new_data
        else:
            # For list, update items that meet condition
            if isinstance(self.data[table_name], list):
                for i, item in enumerate(self.data[table_name]):
                    if condition(item):
                        self.data[table_name][i] = new_data
        
        self._save_to_file()
        return True
    
    def delete_data(self, table_name: str, condition: callable = None, delete_all: bool = False) -> bool:
        """
        Delete data from table.
        
        Args:
            table_name (str): Table name
            condition (callable): Function to determine which data to delete
            delete_all (bool): If True, delete all data in table
            
        Returns:
            bool: True if successful
        """
        if table_name not in self.data:
            print(f"Error: Table '{table_name}' not found.")
            return False
        
        if delete_all:
            if isinstance(self.data[table_name], list):
                self.data[table_name].clear()
            elif isinstance(self.data[table_name], dict):
                self.data[table_name].clear()
            else:
                self.data[table_name] = None
        elif condition:
            if isinstance(self.data[table_name], list):
                self.data[table_name] = [item for item in self.data[table_name] if not condition(item)]
        
        self._save_to_file()
        return True
    
    def get_data(self, table_name: str = None) -> Union[Any, Dict[str, Any]]:
        """
        Retrieve data from table or entire database.
        
        Args:
            table_name (str): Table name (optional)
            
        Returns:
            Data from table or entire database
        """
        if table_name is None:
            return self.data
        
        if table_name not in self.data:
            print(f"Warning: Table '{table_name}' not found.")
            return None
        
        return self.data[table_name]
    
    def show_data(self, table_name: str = None) -> None:
        """
        Display data information in readable format.
        
        Args:
            table_name (str): Table name (optional)
        """
        if table_name is None:
            print("=== Entire Database ===")
            for key, value in self.data.items():
                print(f"Table: {key}")
                print(f"Type: {type(value).__name__}")
                if isinstance(value, (list, dict, str)):
                    print(f"Length: {len(value)}")
                print(f"Data: {value}")
                print("-" * 40)
        else:
            if table_name not in self.data:
                print(f"Table '{table_name}' not found.")
                return
            
            data = self.data[table_name]
            print(f"=== Table: {table_name} ===")
            print(f"Type: {type(data).__name__}")
            if isinstance(data, (list, dict, str)):
                print(f"Length: {len(data)}")
            print(f"Data: {data}")
    
    def list_tables(self) -> List[str]:
        """
        Return list of all table names.
        
        Returns:
            List[str]: List of table names
        """
        return list(self.data.keys())
    
    def table_exists(self, table_name: str) -> bool:
        """
        Check if table exists.
        
        Args:
            table_name (str): Table name
            
        Returns:
            bool: True if table exists
        """
        return table_name in self.data
    
    def get_table_info(self, table_name: str) -> Dict[str, Any]:
        """
        Get detailed information about table.
        
        Args:
            table_name (str): Table name
            
        Returns:
            Dict with table information
        """
        if table_name not in self.data:
            return {"exists": False}
        
        data = self.data[table_name]
        return {
            "exists": True,
            "name": table_name,
            "type": type(data).__name__,
            "length": len(data) if isinstance(data, (list, dict, str)) else 0,
            "data": data
        }
